Assets/MixedRealityToolkit/Utilities/StabilizedRay.cs

```diff
@@ -0,0 +1,99 @@
+﻿using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Utilities
+{
+    /// <summary>
+    /// A ray whose position and direction are stabilized in a way similar to how gaze stabilization
+    /// works in HoloLens.
+    /// 
+    /// The ray uses Anatolie Gavrulic's "DynamicExpDecay" filter to stabilize the ray
+    /// this filter adjusts its smoothing factor based on the velocity of the filtered object
+    /// 
+    /// The formula is
+    ///  Y_smooted += ∆𝑌_𝑟
+    ///  where
+    /// 〖∆𝑌_𝑟=∆𝑌∗〖0.5〗^(∆𝑌/〖Halflife〗)
+    /// 
+    /// In code, LERP(signal, oldValue, POW(0.5, ABS(signal – oldValue) / hl)
+    /// </summary>
+    public class StabilizedRay
+    {
+        /// <summary>
+        /// Half life used for position decay calculations.
+        /// </summary>
+        public float HalfLifePosition { get; } = 0.1f;
+
+        /// <summary>
+        /// Half life used for velocity decay calculations.
+        /// </summary>
+        public float HalfLifeDirection { get; } = 0.1f;
+
+        /// <summary>
+        /// Computed Stabilized position.
+        /// </summary>
+        public Vector3 StabilizedPosition { get; private set; }
+
+        /// <summary>
+        /// Computed stabilized direction.
+        /// </summary>
+        public Vector3 StabilizedDirection { get; private set; }
+
+        private bool isInitialized = false;
+
+        /// <summary>
+        /// HalfLife closer to zero means lerp closer to one.
+        /// </summary>
+        public StabilizedRay(float halfLife)
+        {
+            HalfLifePosition = halfLife;
+            HalfLifeDirection = halfLife;
+        }
+
+        /// <summary>
+        /// Add sample to ray stabilizer.
+        /// </summary>
+        /// <param name="ray">New Sample used to update stabilized ray.</param>
+        public void AddSample(Ray ray)
+        {
+            if (!isInitialized)
+            {
+                StabilizedPosition = ray.origin;
+                StabilizedDirection = ray.direction.normalized;
+                isInitialized = true;
+            }
+            else
+            {
+                StabilizedPosition = DynamicExpDecay(StabilizedPosition, ray.origin, HalfLifePosition);
+                StabilizedDirection = DynamicExpDecay(StabilizedDirection, ray.direction.normalized, HalfLifeDirection);
+            }
+        }
+
+        /// <summary>
+        /// Compute dynamic exponential coefficient.
+        /// </summary>
+        /// <param name="hLife">Half life</param>
+        /// <param name="delta">Distance delta</param>
+        /// <returns></returns>
```

Comment:
PR validation will fail on this empty return

```suggestion
        /// <returns>The dynamic exponential coefficient.</returns>
```

Reply:
Fixed